test/config/utility.cc

```diff
@@ -93,4 +117,84 @@ void ConfigHelper::setSourceAddress(const std::string& address_string) {
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
+
+  auto virtual_host = hcm_config.mutable_route_config()->mutable_virtual_hosts(0);
+  virtual_host->set_domains(0, domains);
+  virtual_host->mutable_routes(0)->mutable_match()->set_prefix(prefix);
+
+  storeHttpConnectionManager(hcm_config);
+}
+
+void ConfigHelper::setBufferLimits(uint32_t upstream_buffer_limit,
+                                   uint32_t downstream_buffer_limit) {
+  RELEASE_ASSERT(bootstrap_.mutable_static_resources()->listeners_size() == 1);
+  auto* listener = bootstrap_.mutable_static_resources()->mutable_listeners(0);
+  listener->mutable_per_connection_buffer_limit_bytes()->set_value(downstream_buffer_limit);
+
+  auto* static_resources = bootstrap_.mutable_static_resources();
+  for (int i = 0; i < bootstrap_.mutable_static_resources()->clusters_size(); ++i) {
+    auto* cluster = static_resources->mutable_clusters(i);
+    cluster->mutable_per_connection_buffer_limit_bytes()->set_value(upstream_buffer_limit);
+  }
+}
+
+void ConfigHelper::addRoute(const std::string& domains, const std::string& prefix,
+                            const std::string& cluster,
+                            envoy::api::v2::VirtualHost::TlsRequirementType type) {
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
+
+  auto virtual_host = hcm_config.mutable_route_config()->add_virtual_hosts();
+  virtual_host->add_domains(domains);
+  virtual_host->add_routes()->mutable_match()->set_prefix(prefix);
+  virtual_host->mutable_routes(0)->mutable_route()->set_cluster(cluster);
+  virtual_host->set_require_tls(type);
+
+  storeHttpConnectionManager(hcm_config);
+}
+
+void ConfigHelper::addFilter(const std::string& config) {
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
+
+  auto* filter_list_back = hcm_config.add_http_filters();
+  const std::string json = Json::Factory::loadFromYamlString(config)->asJsonString();
+  MessageUtil::loadFromJson(json, *filter_list_back);
+
+  // Now move it to the front.
+  for (int i = hcm_config.http_filters_size() - 1; i > 0; --i) {
+    hcm_config.mutable_http_filters()->SwapElements(i, i - 1);
+  }
+
+  storeHttpConnectionManager(hcm_config);
+}
+
+void ConfigHelper::loadHttpConnectionManager(envoy::api::v2::filter::HttpConnectionManager& hcm) {
+  auto* listener = bootstrap_.mutable_static_resources()->mutable_listeners(0);
+  auto* filter_chain = listener->mutable_filter_chains(0);
+  auto* hcm_config_struct = filter_chain->mutable_filters(0)->mutable_config();
+  MessageUtil::jsonConvert(*hcm_config_struct, hcm);
+}
+
+void ConfigHelper::storeHttpConnectionManager(
+    const envoy::api::v2::filter::HttpConnectionManager& hcm) {
+  auto* listener = bootstrap_.mutable_static_resources()->mutable_listeners(0);
```

Comment:
Do you want to build this API to handle more than 1 listener eventually?

Reply:
I figured we'd extend as needed.

I've made it through about half the tests and there's only 1 which requires 2 listeners, which are the forward-proxy tests which set up one listener which handles abs urls, one which does not, and verifies some behavior is the same.

I found it easy to clone the listener and call the edit function to alter the first.

Reply:
Oh thought I like getting rid of ExecuteActions even more if we can swing it. 
Not in the same CL however :-P